checker/src/org/checkerframework/checker/index/searchindex/SearchIndexAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,106 @@
+        super(checker);
+        UNKNOWN = AnnotationUtils.fromClass(elements, SearchIndexUnknown.class);
+        BOTTOM = AnnotationUtils.fromClass(elements, SearchIndexBottom.class);
+        this.postInit();
+    }
+
+    /**
+     * Provides a way to query the Constant Value Checker, which computes the values of expressions
+     * known at compile time (constant prop + folding).
+     */
+    ValueAnnotatedTypeFactory getValueAnnotatedTypeFactory() {
+        return getTypeFactoryOfSubchecker(ValueChecker.class);
+    }
+
+    @Override
+    protected Set<Class<? extends Annotation>> createSupportedTypeQualifiers() {
+        return new LinkedHashSet<>(
+                Arrays.asList(
+                        SearchIndex.class,
+                        SearchIndexBottom.class,
+                        SearchIndexUnknown.class,
+                        NegativeIndexFor.class));
+    }
+
+    @Override
+    public QualifierHierarchy createQualifierHierarchy(
+            MultiGraphQualifierHierarchy.MultiGraphFactory factory) {
+        return new SearchIndexQualifierHierarchy(factory);
+    }
+
+    private final class SearchIndexQualifierHierarchy extends MultiGraphQualifierHierarchy {
+
+        public SearchIndexQualifierHierarchy(
+                MultiGraphQualifierHierarchy.MultiGraphFactory factory) {
+            super(factory);
+        }
+
+        @Override
+        public boolean isSubtype(AnnotationMirror subAnno, AnnotationMirror superAnno) {
+            if (AnnotationUtils.areSameByClass(superAnno, SearchIndexUnknown.class)) {
+                return true;
+            }
+            if (AnnotationUtils.areSameByClass(subAnno, SearchIndexBottom.class)) {
+                return true;
+            }
+            if (AnnotationUtils.areSameByClass(subAnno, SearchIndexUnknown.class)) {
+                return false;
+            }
+            if (AnnotationUtils.areSameByClass(superAnno, SearchIndexBottom.class)) {
+                return false;
+            }
+            if (AnnotationUtils.areSameByClass(subAnno, SearchIndex.class)) {
+                if (AnnotationUtils.areSameByClass(superAnno, NegativeIndexFor.class)) {
+                    return false;
+                }
+                return IndexUtil.getValueOfAnnotationWithStringArgument(superAnno)
+                        .containsAll(IndexUtil.getValueOfAnnotationWithStringArgument(subAnno));
+            }
+            if (AnnotationUtils.areSameByClass(subAnno, NegativeIndexFor.class)) {
+                if (AnnotationUtils.areSameByClass(superAnno, SearchIndex.class)) {
+                    return true;
+                }
+                return IndexUtil.getValueOfAnnotationWithStringArgument(superAnno)
+                        .containsAll(IndexUtil.getValueOfAnnotationWithStringArgument(subAnno));
+            }
+            return false; // dead code
+        }
+    }
+
+    AnnotationMirror createNegativeIndexFor(List<String> arrays) {
```

Comment:
Please write a brief comment for each method and field.  This one is pretty obvious, but a one-line comment will still help.